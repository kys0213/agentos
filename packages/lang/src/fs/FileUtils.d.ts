import fs from 'fs/promises';
import { Result } from '../utils/safeZone';
export interface FileOptions {
    /** 파일이 없을 경우 생성 여부 (default: false) */
    createIfNotExists?: boolean;
    /** 디렉터리가 없을 경우 생성 여부 (default: true) */
    ensureDir?: boolean;
    /** 파일 인코딩 (default: 'utf-8') */
    encoding?: BufferEncoding;
}
export interface DirectoryOptions {
    /** 재귀적으로 디렉터리 생성 여부 (default: true) */
    recursive?: boolean;
}
/**
 * 기본 파일 시스템 유틸리티
 * 모든 파일 I/O 작업의 기반이 되는 클래스
 */
export declare class FileUtils {
    /**
     * 파일 또는 디렉터리 존재 확인
     */
    static exists(filePath: string): Promise<boolean>;
    /**
     * 디렉터리 생성 (재귀적으로)
     */
    static ensureDir(dirPath: string, options?: DirectoryOptions): Promise<Result<void>>;
    /**
     * 안전한 파일 읽기
     */
    static readSafe(filePath: string, options?: FileOptions): Promise<Result<string>>;
    /**
     * 안전한 파일 쓰기
     */
    static writeSafe(filePath: string, content: string, options?: FileOptions): Promise<Result<void>>;
    /**
     * 안전한 파일 추가 (append)
     */
    static appendSafe(filePath: string, content: string, options?: FileOptions): Promise<Result<void>>;
    /**
     * 파일 삭제
     */
    static remove(filePath: string): Promise<Result<void>>;
    /**
     * 파일 통계 정보 가져오기
     */
    static stat(filePath: string): Promise<Result<Awaited<ReturnType<typeof fs.stat>>>>;
    /**
     * 디렉터리 내용 읽기
     */
    static readDir(dirPath: string): Promise<Result<string[]>>;
    /**
     * 파일 복사
     */
    static copy(srcPath: string, destPath: string, options?: FileOptions): Promise<Result<void>>;
}
//# sourceMappingURL=FileUtils.d.ts.map