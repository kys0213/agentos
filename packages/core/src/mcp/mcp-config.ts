export interface StdioMcpConfig {
  name: string;
  version: string;
  command: string;
  args?: string[];
  env?: Record<string, string>;
  cwd?: string;
  network?: {
    timeout?: number;
    maxTotalTimeout?: number;
  };
}
