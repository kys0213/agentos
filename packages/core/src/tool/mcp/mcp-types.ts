/**
 * Core MCP Usage Tracking Types
 * 
 * GUI의 MCPTool/ToolUsageLog 타입을 Core로 통합하여
 * 모든 앱에서 일관된 MCP 사용량 데이터 활용이 가능하도록 합니다.
 */

/**
 * MCP 도구의 메타데이터 정보
 */
export interface McpToolMetadata {
  /** 고유 식별자 */
  id: string;
  /** 도구 이름 */
  name: string;
  /** 도구 설명 */
  description: string;
  /** 도구 버전 */
  version: string;
  /** 도구 카테고리 (옵셔널) */
  category?: string;
  /** 도구 제공자 (옵셔널) */
  provider?: string;
  /** 연결 엔드포인트 (옵셔널) */
  endpoint?: string;
  /** 도구 권한 목록 */
  permissions: string[];
  /** 연결 상태 */
  status: McpConnectionStatus;
  /** 마지막 사용 시간 (옵셔널) */
  lastUsedAt?: Date;
  /** 총 사용 횟수 */
  usageCount: number;
  /** 도구별 설정 (옵셔널) */
  config?: Record<string, any>;
}

/**
 * MCP 도구 사용 로그
 */
export interface McpUsageLog {
  /** 로그 고유 식별자 */
  id: string;
  /** 사용된 도구 ID */
  toolId: string;
  /** 사용된 도구 이름 */
  toolName: string;
  /** 사용한 에이전트 ID (옵셔널) */
  agentId?: string;
  /** 사용한 에이전트 이름 (옵셔널) */
  agentName?: string;
  /** 실행된 액션 */
  action: string;
  /** 사용 시간 */
  timestamp: Date;
  /** 실행 시간 (밀리초) */
  duration: number;
  /** 실행 결과 상태 */
  status: McpUsageStatus;
  /** 입력 파라미터 (옵셔널) */
  parameters?: Record<string, any>;
  /** 에러 메시지 (옵셔널) */
  error?: string;
  /** 실행 결과 (옵셔널) */
  result?: string;
}

/**
 * MCP 연결 상태
 */
export type McpConnectionStatus = 'connected' | 'disconnected' | 'error' | 'pending';

/**
 * MCP 사용 결과 상태
 */
export type McpUsageStatus = 'success' | 'error' | 'timeout';

/**
 * MCP 사용량 추적기 인터페이스
 */
export interface McpUsageTracker {
  /**
   * 사용량 로그 기록
   */
  trackUsage(log: Omit<McpUsageLog, 'id' | 'timestamp'>): void;
  
  /**
   * 사용량 로그 조회
   * @param toolId 특정 도구 ID (옵셔널, 미지정 시 전체 조회)
   */
  getUsageLogs(toolId?: string): McpUsageLog[];
  
  /**
   * 사용량 통계 조회
   * @param toolId 특정 도구 ID (옵셔널, 미지정 시 전체 통계)
   */
  getUsageStats(toolId?: string): McpUsageStats;
  
  /**
   * 오래된 로그 삭제
   * @param olderThan 기준 날짜 (옵셔널, 미지정 시 전체 삭제)
   */
  clearLogs(olderThan?: Date): void;
}

/**
 * MCP 사용량 통계
 */
export interface McpUsageStats {
  /** 총 사용 횟수 */
  totalUsage: number;
  /** 성공률 (0-1) */
  successRate: number;
  /** 평균 실행 시간 (밀리초) */
  averageDuration: number;
  /** 마지막 사용 시간 (옵셔널) */
  lastUsedAt?: Date;
  /** 에러 발생 횟수 */
  errorCount: number;
}