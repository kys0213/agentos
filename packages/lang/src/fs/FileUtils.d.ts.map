{"version":3,"file":"FileUtils.d.ts","sourceRoot":"","sources":["FileUtils.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,MAAM,aAAa,CAAC;AAE7B,OAAO,EAAE,MAAM,EAAE,MAAM,mBAAmB,CAAC;AAE3C,MAAM,WAAW,WAAW;IAC1B,uCAAuC;IACvC,iBAAiB,CAAC,EAAE,OAAO,CAAC;IAC5B,wCAAwC;IACxC,SAAS,CAAC,EAAE,OAAO,CAAC;IACpB,gCAAgC;IAChC,QAAQ,CAAC,EAAE,cAAc,CAAC;CAC3B;AAED,MAAM,WAAW,gBAAgB;IAC/B,uCAAuC;IACvC,SAAS,CAAC,EAAE,OAAO,CAAC;CACrB;AAED;;;GAGG;AACH,qBAAa,SAAS;IACpB;;OAEG;WACU,MAAM,CAAC,QAAQ,EAAE,MAAM,GAAG,OAAO,CAAC,OAAO,CAAC;IASvD;;OAEG;WACU,SAAS,CAAC,OAAO,EAAE,MAAM,EAAE,OAAO,GAAE,gBAAqB,GAAG,OAAO,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC;IAW9F;;OAEG;WACU,QAAQ,CAAC,QAAQ,EAAE,MAAM,EAAE,OAAO,GAAE,WAAgB,GAAG,OAAO,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC;IA4B3F;;OAEG;WACU,SAAS,CACpB,QAAQ,EAAE,MAAM,EAChB,OAAO,EAAE,MAAM,EACf,OAAO,GAAE,WAAgB,GACxB,OAAO,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC;IAoBxB;;OAEG;WACU,UAAU,CACrB,QAAQ,EAAE,MAAM,EAChB,OAAO,EAAE,MAAM,EACf,OAAO,GAAE,WAAgB,GACxB,OAAO,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC;IA4BxB;;OAEG;WACU,MAAM,CAAC,QAAQ,EAAE,MAAM,GAAG,OAAO,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC;IAW5D;;OAEG;WACU,IAAI,CAAC,QAAQ,EAAE,MAAM,GAAG,OAAO,CAAC,MAAM,CAAC,OAAO,CAAC,UAAU,CAAC,OAAO,EAAE,CAAC,IAAI,CAAC,CAAC,CAAC,CAAC;IASzF;;OAEG;WACU,OAAO,CAAC,OAAO,EAAE,MAAM,GAAG,OAAO,CAAC,MAAM,CAAC,MAAM,EAAE,CAAC,CAAC;IAShE;;OAEG;WACU,IAAI,CACf,OAAO,EAAE,MAAM,EACf,QAAQ,EAAE,MAAM,EAChB,OAAO,GAAE,WAAgB,GACxB,OAAO,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC;CAmBzB"}