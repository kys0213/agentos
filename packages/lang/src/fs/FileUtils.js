"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileUtils = void 0;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
/**
 * 기본 파일 시스템 유틸리티
 * 모든 파일 I/O 작업의 기반이 되는 클래스
 */
class FileUtils {
    /**
     * 파일 또는 디렉터리 존재 확인
     */
    static async exists(filePath) {
        try {
            await promises_1.default.access(filePath);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * 디렉터리 생성 (재귀적으로)
     */
    static async ensureDir(dirPath, options = {}) {
        const { recursive = true } = options;
        try {
            await promises_1.default.mkdir(dirPath, { recursive });
            return { success: true, result: undefined };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 안전한 파일 읽기
     */
    static async readSafe(filePath, options = {}) {
        const { encoding = 'utf-8', createIfNotExists = false, ensureDir = true } = options;
        try {
            // 디렉터리 확인 및 생성
            if (ensureDir) {
                const dirPath = path_1.default.dirname(filePath);
                const dirResult = await this.ensureDir(dirPath);
                if (!dirResult.success) {
                    return { success: false, reason: dirResult.reason };
                }
            }
            // 파일이 없고 생성 옵션이 true인 경우
            if (!(await this.exists(filePath)) && createIfNotExists) {
                const writeResult = await this.writeSafe(filePath, '', options);
                if (!writeResult.success) {
                    return { success: false, reason: writeResult.reason };
                }
            }
            const content = await promises_1.default.readFile(filePath, encoding);
            return { success: true, result: content };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 안전한 파일 쓰기
     */
    static async writeSafe(filePath, content, options = {}) {
        const { encoding = 'utf-8', ensureDir = true } = options;
        try {
            // 디렉터리 확인 및 생성
            if (ensureDir) {
                const dirPath = path_1.default.dirname(filePath);
                const dirResult = await this.ensureDir(dirPath);
                if (!dirResult.success) {
                    return { success: false, reason: dirResult.reason };
                }
            }
            await promises_1.default.writeFile(filePath, content, encoding);
            return { success: true, result: undefined };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 안전한 파일 추가 (append)
     */
    static async appendSafe(filePath, content, options = {}) {
        const { encoding = 'utf-8', ensureDir = true, createIfNotExists = true } = options;
        try {
            // 디렉터리 확인 및 생성
            if (ensureDir) {
                const dirPath = path_1.default.dirname(filePath);
                const dirResult = await this.ensureDir(dirPath);
                if (!dirResult.success) {
                    return { success: false, reason: dirResult.reason };
                }
            }
            // 파일이 없고 생성 옵션이 true인 경우
            if (!(await this.exists(filePath)) && createIfNotExists) {
                const writeResult = await this.writeSafe(filePath, '', options);
                if (!writeResult.success) {
                    return { success: false, reason: writeResult.reason };
                }
            }
            await promises_1.default.appendFile(filePath, content, encoding);
            return { success: true, result: undefined };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 파일 삭제
     */
    static async remove(filePath) {
        try {
            if (await this.exists(filePath)) {
                await promises_1.default.unlink(filePath);
            }
            return { success: true, result: undefined };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 파일 통계 정보 가져오기
     */
    static async stat(filePath) {
        try {
            const stats = await promises_1.default.stat(filePath);
            return { success: true, result: stats };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 디렉터리 내용 읽기
     */
    static async readDir(dirPath) {
        try {
            const entries = await promises_1.default.readdir(dirPath);
            return { success: true, result: entries };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
    /**
     * 파일 복사
     */
    static async copy(srcPath, destPath, options = {}) {
        const { ensureDir = true } = options;
        try {
            // 대상 디렉터리 확인 및 생성
            if (ensureDir) {
                const dirPath = path_1.default.dirname(destPath);
                const dirResult = await this.ensureDir(dirPath);
                if (!dirResult.success) {
                    return { success: false, reason: dirResult.reason };
                }
            }
            await promises_1.default.copyFile(srcPath, destPath);
            return { success: true, result: undefined };
        }
        catch (error) {
            return { success: false, reason: error };
        }
    }
}
exports.FileUtils = FileUtils;
//# sourceMappingURL=FileUtils.js.map