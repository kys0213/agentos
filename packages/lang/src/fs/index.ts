export * from './FileUtils';
export * from './JsonFileHandler';
export * from './JsonLFileHandler';
