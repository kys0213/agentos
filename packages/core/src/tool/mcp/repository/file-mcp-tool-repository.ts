import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { CursorPagination, CursorPaginationResult } from '../../../common/pagination/cursor-pagination';
import { paginateByCursor } from '../../../common/pagination/paginate';
import { SimpleEventEmitter } from '../../../common/event/simple-event-emitter';
import { McpToolMetadata, McpConnectionStatus } from '../mcp-types';
import { McpConfig } from '../mcp-config';
import {
  McpToolRepository,
  McpToolSearchQuery,
  McpToolRepositoryEvent,
  McpToolRepositoryEventHandler,
  McpToolRepositoryEventPayload,
  Unsubscribe
} from './mcp-tool-repository';

/**
 * 파일 기반 MCP 도구 Repository 구현
 * 
 * JSON 파일을 사용하여 도구 메타데이터를 영속화합니다.
 * 이벤트 기반 반응성을 위해 SimpleEventEmitter를 사용합니다.
 */
export class FileMcpToolRepository implements McpToolRepository {
  private readonly eventEmitter = new SimpleEventEmitter<{
    changed: McpToolRepositoryEventPayload;
    deleted: McpToolRepositoryEventPayload;
    statusChanged: McpToolRepositoryEventPayload;
  }>();

  private cache: Map<string, McpToolMetadata> = new Map();
  private initialized = false;

  constructor(
    private readonly storagePath: string,
    private readonly options: {
      /** 파일 변경 감지 활성화 여부 */
      enableWatching?: boolean;
      /** 캐시 사용 여부 */
      enableCaching?: boolean;
    } = {}
  ) {
    this.options = {
      enableWatching: true,
      enableCaching: true,
      ...options
    };
  }

  async get(id: string): Promise<McpToolMetadata | null> {
    await this.ensureInitialized();

    if (this.options.enableCaching && this.cache.has(id)) {
      return this.cache.get(id)!;
    }

    const tools = await this.loadAll();
    const tool = tools.find(t => t.id === id);

    if (tool && this.options.enableCaching) {
      this.cache.set(id, tool);
    }

    return tool || null;
  }

  async list(pagination?: CursorPagination): Promise<CursorPaginationResult<McpToolMetadata>> {
    await this.ensureInitialized();
    const tools = await this.loadAll();

    return paginateByCursor(tools, pagination);
  }

  async search(
    query: McpToolSearchQuery,
    pagination?: CursorPagination
  ): Promise<CursorPaginationResult<McpToolMetadata>> {
    await this.ensureInitialized();
    const tools = await this.loadAll();

    const filtered = tools.filter(tool => {
      // 카테고리 필터
      if (query.category && tool.category !== query.category) {
        return false;
      }

      // 상태 필터
      if (query.status && tool.status !== query.status) {
        return false;
      }

      // 제공자 필터
      if (query.provider && tool.provider !== query.provider) {
        return false;
      }

      // 키워드 검색
      if (query.keywords && query.keywords.length > 0) {
        const searchText = `${tool.name} ${tool.description}`.toLowerCase();
        const hasMatchingKeyword = query.keywords.some(keyword =>
          searchText.includes(keyword.toLowerCase())
        );
        if (!hasMatchingKeyword) {
          return false;
        }
      }

      return true;
    });

    return paginateByCursor(filtered, pagination);
  }

  async create(config: McpConfig): Promise<McpToolMetadata> {
    await this.ensureInitialized();

    const metadata: McpToolMetadata = {
      id: this.generateToolId(config),
      name: config.name,
      description: `MCP Tool: ${config.name}`,
      version: config.version,
      category: this.inferCategory(config),
      provider: this.inferProvider(config),
      endpoint: this.extractEndpoint(config),
      permissions: [],
      status: 'disconnected',
      usageCount: 0,
      config: this.sanitizeConfig(config),
    };

    const tools = await this.loadAll();
    tools.push(metadata);
    await this.saveAll(tools);

    if (this.options.enableCaching) {
      this.cache.set(metadata.id, metadata);
    }

    this.eventEmitter.emit('changed', { id: metadata.id, metadata });

    return metadata;
  }

  async update(
    id: string,
    patch: Partial<McpToolMetadata>,
    options?: { expectedVersion?: string }
  ): Promise<McpToolMetadata> {
    await this.ensureInitialized();

    const tools = await this.loadAll();
    const index = tools.findIndex(t => t.id === id);

    if (index === -1) {
      throw new Error(`MCP tool not found: ${id}`);
    }

    const existing = tools[index];

    // 낙관적 동시성 제어
    if (options?.expectedVersion && existing.version !== options.expectedVersion) {
      throw new Error(
        `Version conflict for tool ${id}. Expected: ${options.expectedVersion}, Current: ${existing.version}`
      );
    }

    const previousStatus = existing.status;
    const updated: McpToolMetadata = {
      ...existing,
      ...patch,
      id, // ID는 변경 불가
      version: this.generateVersion(), // 새 버전 할당
      lastUsedAt: patch.usageCount !== existing.usageCount ? new Date() : existing.lastUsedAt,
    };

    tools[index] = updated;
    await this.saveAll(tools);

    if (this.options.enableCaching) {
      this.cache.set(id, updated);
    }

    // 이벤트 발행
    this.eventEmitter.emit('changed', { id, metadata: updated });
    
    if (previousStatus !== updated.status) {
      this.eventEmitter.emit('statusChanged', {
        id,
        metadata: updated,
        previousStatus
      });
    }

    return updated;
  }

  async delete(id: string): Promise<void> {
    await this.ensureInitialized();

    const tools = await this.loadAll();
    const index = tools.findIndex(t => t.id === id);

    if (index === -1) {
      throw new Error(`MCP tool not found: ${id}`);
    }

    tools.splice(index, 1);
    await this.saveAll(tools);

    if (this.options.enableCaching) {
      this.cache.delete(id);
    }

    this.eventEmitter.emit('deleted', { id });
  }

  on(
    event: McpToolRepositoryEvent,
    handler: McpToolRepositoryEventHandler
  ): Unsubscribe {
    return this.eventEmitter.on(event, handler);
  }

  /**
   * Repository 초기화 (저장소 디렉토리 생성 등)
   */
  private async ensureInitialized(): Promise<void> {
    if (this.initialized) return;

    try {
      await fs.mkdir(path.dirname(this.storagePath), { recursive: true });

      // 파일이 존재하지 않으면 빈 배열로 초기화
      try {
        await fs.access(this.storagePath);
      } catch {
        await this.saveAll([]);
      }

      this.initialized = true;
    } catch (error) {
      throw new Error(`Failed to initialize MCP tool repository: ${error}`);
    }
  }

  /**
   * 모든 도구 메타데이터 로드
   */
  private async loadAll(): Promise<McpToolMetadata[]> {
    try {
      const content = await fs.readFile(this.storagePath, 'utf-8');
      return JSON.parse(content) as McpToolMetadata[];
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        return [];
      }
      throw new Error(`Failed to load MCP tools: ${error}`);
    }
  }

  /**
   * 모든 도구 메타데이터 저장
   */
  private async saveAll(tools: McpToolMetadata[]): Promise<void> {
    try {
      const content = JSON.stringify(tools, null, 2);
      await fs.writeFile(this.storagePath, content, 'utf-8');
    } catch (error) {
      throw new Error(`Failed to save MCP tools: ${error}`);
    }
  }

  /**
   * 도구 ID 생성
   */
  private generateToolId(config: McpConfig): string {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 8);
    return `mcp_${config.name.replace(/[^a-zA-Z0-9]/g, '_')}_${timestamp}_${random}`;
  }

  /**
   * 버전 문자열 생성 (ETag 스타일)
   */
  private generateVersion(): string {
    return `v${Date.now()}_${Math.random().toString(36).substring(2, 8)}`;
  }

  /**
   * 설정으로부터 카테고리 추론
   */
  private inferCategory(config: McpConfig): string {
    const name = config.name.toLowerCase();

    if (name.includes('search') || name.includes('web')) return 'search';
    if (name.includes('code') || name.includes('dev')) return 'development';
    if (name.includes('data') || name.includes('db')) return 'data';
    if (name.includes('file') || name.includes('doc')) return 'productivity';
    if (name.includes('api') || name.includes('service')) return 'api';

    return 'general';
  }

  /**
   * 설정으로부터 제공자 추론
   */
  private inferProvider(config: McpConfig): string {
    if (config.type === 'stdio') {
      return 'Local Process';
    }

    if ('url' in config && config.url) {
      try {
        const url = new URL(config.url);
        return url.hostname;
      } catch {
        return 'External Service';
      }
    }

    return 'Unknown';
  }

  /**
   * 설정으로부터 엔드포인트 추출
   */
  private extractEndpoint(config: McpConfig): string | undefined {
    if (config.type === 'stdio') {
      return `${config.command} ${(config.args || []).join(' ')}`.trim();
    }

    if ('url' in config) {
      return config.url;
    }

    return undefined;
  }

  /**
   * 민감한 정보 제거한 설정 저장
   */
  private sanitizeConfig(config: McpConfig): Record<string, any> {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { env, headers, ...sanitized } = config as any;
    
    return {
      ...sanitized,
      // 환경변수나 헤더에서 민감한 정보 제거
      ...(env && { env: this.sanitizeEnv(env) }),
      ...(headers && { headers: this.sanitizeHeaders(headers) }),
    };
  }

  /**
   * 환경변수에서 민감한 값 마스킹
   */
  private sanitizeEnv(env: Record<string, string>): Record<string, string> {
    const sensitiveKeys = ['api_key', 'secret', 'token', 'password', 'key'];
    const sanitized: Record<string, string> = {};

    for (const [key, value] of Object.entries(env)) {
      const isSensitive = sensitiveKeys.some(sensitive =>
        key.toLowerCase().includes(sensitive)
      );

      sanitized[key] = isSensitive ? '***masked***' : value;
    }

    return sanitized;
  }

  /**
   * 헤더에서 민감한 값 마스킹
   */
  private sanitizeHeaders(headers: Record<string, string>): Record<string, string> {
    const sensitiveHeaders = ['authorization', 'x-api-key', 'authentication'];
    const sanitized: Record<string, string> = {};

    for (const [key, value] of Object.entries(headers)) {
      const isSensitive = sensitiveHeaders.some(sensitive =>
        key.toLowerCase().includes(sensitive)
      );

      sanitized[key] = isSensitive ? '***masked***' : value;
    }

    return sanitized;
  }
}