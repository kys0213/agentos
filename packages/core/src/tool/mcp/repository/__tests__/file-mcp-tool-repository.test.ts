import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import * as os from 'node:os';
import { FileMcpToolRepository } from '../file-mcp-tool-repository';
import { McpConfig } from '../../mcp-config';
import { McpToolMetadata } from '../../mcp-types';

describe('FileMcpToolRepository', () => {
  let repository: FileMcpToolRepository;
  let tempDir: string;
  let storagePath: string;

  beforeEach(async () => {
    // 임시 디렉토리 생성
    tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'mcp-repo-test-'));
    storagePath = path.join(tempDir, 'mcp-tools.json');

    repository = new FileMcpToolRepository(storagePath, {
      enableCaching: true,
      enableWatching: false, // 테스트에서는 파일 감시 비활성화
    });
  });

  afterEach(async () => {
    // 임시 디렉토리 정리
    try {
      await fs.rm(tempDir, { recursive: true, force: true });
    } catch {
      // 정리 실패는 무시
    }
  });

  describe('create', () => {
    it('should create a new MCP tool metadata', async () => {
      const config: McpConfig = {
        type: 'stdio',
        name: 'test-tool',
        version: '1.0.0',
        command: 'node',
        args: ['script.js'],
        env: { API_KEY: 'secret123' },
      };

      const metadata = await repository.create(config);

      expect(metadata).toMatchObject({
        name: 'test-tool',
        version: '1.0.0',
        category: 'general',
        provider: 'Local Process',
        status: 'disconnected',
        usageCount: 0,
      });

      expect(metadata.id).toMatch(/^mcp_test_tool_\d+_[a-z0-9]+$/);
      expect(metadata.endpoint).toBe('node script.js');
      expect(metadata.config?.env?.API_KEY).toBe('***masked***'); // 민감한 정보 마스킹
    });

    it('should infer category from tool name', async () => {
      const webConfig: McpConfig = {
        type: 'streamableHttp',
        name: 'web-search-tool',
        version: '1.0.0',
        url: 'https://api.example.com',
      };

      const metadata = await repository.create(webConfig);
      expect(metadata.category).toBe('search');
    });

    it('should emit changed event on creation', async () => {
      const config: McpConfig = {
        type: 'stdio',
        name: 'test-tool',
        version: '1.0.0',
        command: 'node',
      };

      const eventPromise = new Promise((resolve) => {
        repository.on?.('changed', resolve);
      });

      const metadata = await repository.create(config);
      const event = await eventPromise;

      expect(event).toEqual({
        id: metadata.id,
        metadata,
      });
    });
  });

  describe('get', () => {
    it('should retrieve existing tool by id', async () => {
      const config: McpConfig = {
        type: 'stdio',
        name: 'test-tool',
        version: '1.0.0',
        command: 'node',
      };

      const created = await repository.create(config);
      const retrieved = await repository.get(created.id);

      expect(retrieved).toEqual(created);
    });

    it('should return null for non-existent tool', async () => {
      const result = await repository.get('non-existent-id');
      expect(result).toBeNull();
    });
  });

  describe('list', () => {
    it('should return empty list initially', async () => {
      const result = await repository.list();

      expect(result.items).toHaveLength(0);
      expect(result.hasMore).toBe(false);
      expect(result.nextCursor).toBe('');
    });

    it('should return all tools', async () => {
      const configs: McpConfig[] = [
        { type: 'stdio', name: 'tool1', version: '1.0.0', command: 'node' },
        { type: 'stdio', name: 'tool2', version: '1.0.0', command: 'python' },
      ];

      for (const config of configs) {
        await repository.create(config);
      }

      const result = await repository.list();

      expect(result.items).toHaveLength(2);
      expect(result.items.map((t) => t.name)).toContain('tool1');
      expect(result.items.map((t) => t.name)).toContain('tool2');
    });

    it('should support pagination', async () => {
      // 3개 도구 생성
      for (let i = 1; i <= 3; i++) {
        await repository.create({
          type: 'stdio',
          name: `tool${i}`,
          version: '1.0.0',
          command: 'node',
        });
      }

      // 첫 페이지 (limit: 2)
      const firstPage = await repository.list({ limit: 2, cursor: '', direction: 'forward' });
      expect(firstPage.items).toHaveLength(2);
      expect(firstPage.hasMore).toBe(true);
      expect(firstPage.nextCursor).toBeDefined();

      // 두 번째 페이지
      const secondPage = await repository.list({
        limit: 2,
        cursor: firstPage.nextCursor,
        direction: 'forward',
      });
      expect(secondPage.items).toHaveLength(1);
      expect(secondPage.hasMore).toBe(false);
    });
  });

  describe('search', () => {
    beforeEach(async () => {
      const configs: McpConfig[] = [
        {
          type: 'stdio',
          name: 'web-search',
          version: '1.0.0',
          command: 'node',
          args: ['web-search.js'],
        },
        {
          type: 'streamableHttp',
          name: 'code-executor',
          version: '1.0.0',
          url: 'https://api.codeexec.com',
        },
        {
          type: 'stdio',
          name: 'file-manager',
          version: '1.0.0',
          command: 'python',
          args: ['file_manager.py'],
        },
      ];

      for (const config of configs) {
        await repository.create(config);
      }
    });

    it('should search by category', async () => {
      const result = await repository.search({ category: 'search' });

      expect(result.items).toHaveLength(1);
      expect(result.items[0].name).toBe('web-search');
    });

    it('should search by keywords', async () => {
      const result = await repository.search({ keywords: ['code'] });

      expect(result.items).toHaveLength(1);
      expect(result.items[0].name).toBe('code-executor');
    });

    it('should search by multiple criteria', async () => {
      // provider가 'Local Process'이면서 키워드에 'file'이 포함된 도구
      const result = await repository.search({
        provider: 'Local Process',
        keywords: ['file'],
      });

      expect(result.items).toHaveLength(1);
      expect(result.items[0].name).toBe('file-manager');
    });

    it('should return empty result for no matches', async () => {
      const result = await repository.search({ keywords: ['nonexistent'] });
      expect(result.items).toHaveLength(0);
    });
  });

  describe('update', () => {
    let toolId: string;

    beforeEach(async () => {
      const config: McpConfig = {
        type: 'stdio',
        name: 'test-tool',
        version: '1.0.0',
        command: 'node',
      };

      const metadata = await repository.create(config);
      toolId = metadata.id;
    });

    it('should update tool metadata', async () => {
      const patch = {
        status: 'connected' as const,
        usageCount: 5,
      };

      const updated = await repository.update(toolId, patch);

      expect(updated.status).toBe('connected');
      expect(updated.usageCount).toBe(5);
      expect(updated.lastUsedAt).toBeInstanceOf(Date);
      expect(updated.version).not.toBe('1.0.0'); // 새 버전 할당됨
    });

    it('should emit statusChanged event on status update', async () => {
      const eventPromise = new Promise((resolve) => {
        repository.on?.('statusChanged', resolve);
      });

      await repository.update(toolId, { status: 'connected' });
      const event = await eventPromise;

      expect(event).toMatchObject({
        id: toolId,
        previousStatus: 'disconnected',
      });
    });

    it('should handle version conflicts', async () => {
      const wrongVersion = 'wrong-version';

      await expect(
        repository.update(
          toolId,
          { status: 'connected' },
          {
            expectedVersion: wrongVersion,
          }
        )
      ).rejects.toThrow(/Version conflict/);
    });

    it('should throw error for non-existent tool', async () => {
      await expect(repository.update('non-existent', { status: 'connected' })).rejects.toThrow(
        /not found/
      );
    });
  });

  describe('delete', () => {
    let toolId: string;

    beforeEach(async () => {
      const config: McpConfig = {
        type: 'stdio',
        name: 'test-tool',
        version: '1.0.0',
        command: 'node',
      };

      const metadata = await repository.create(config);
      toolId = metadata.id;
    });

    it('should delete existing tool', async () => {
      await repository.delete(toolId);

      const result = await repository.get(toolId);
      expect(result).toBeNull();
    });

    it('should emit deleted event', async () => {
      const eventPromise = new Promise((resolve) => {
        repository.on?.('deleted', resolve);
      });

      await repository.delete(toolId);
      const event = await eventPromise;

      expect(event).toEqual({ id: toolId });
    });

    it('should throw error for non-existent tool', async () => {
      await expect(repository.delete('non-existent')).rejects.toThrow(/not found/);
    });
  });

  describe('persistence', () => {
    it('should persist data across repository instances', async () => {
      const config: McpConfig = {
        type: 'stdio',
        name: 'persistent-tool',
        version: '1.0.0',
        command: 'node',
      };

      // 첫 번째 인스턴스로 데이터 생성
      const metadata = await repository.create(config);

      // 새 인스턴스 생성
      const newRepository = new FileMcpToolRepository(storagePath);
      const retrieved = await newRepository.get(metadata.id);

      expect(retrieved).toEqual(metadata);
    });
  });
});
