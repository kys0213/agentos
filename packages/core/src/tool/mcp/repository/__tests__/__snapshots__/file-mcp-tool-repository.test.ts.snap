// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileMcpToolRepository create should create a new MCP tool metadata 1`] = `
{
  "args": [
    "script.js",
  ],
  "command": "node",
  "cwd": undefined,
  "env": {
    "API_KEY": "***masked***",
  },
  "name": "test-tool",
  "network": undefined,
  "type": "stdio",
  "version": "1.0.0",
}
`;
